src/formats/XYZ.cpp

```diff
@@ -252,6 +304,115 @@ std::string write_extended_comment_line(const Frame& frame) {
+        if (!is_valid_property_name(property.first)) {
+            warning(
+                "Extended XYZ", "'{}' is not a valid property name for extended "
+                "XYZ, is will not be saved",
+                property.first, 0
+            );
+            partially_defined_already_warned.insert(property.first);
+            continue;
+        }
+
+        if (property.second.kind() == Property::STRING) {
+            if (should_be_quoted(property.second.as_string())) {
+                warning(
+                    "Extended XYZ", "string value for property '{}' on atom {} "
+                    "can not be be saved as an atomic property",
+                    property.first, 0
+                );
+                continue;
+            }
+        }
+
+        all_properties.emplace(property.first, property.second.kind());
+    }
+
+    for (size_t i=1; i<frame.size(); i++) {
+        const auto& atom = frame[i];
+
+        // warn for properties defined on this atom but not on others
+        for (const auto& property: atom.properties()) {
+            if (all_properties.count(property.first) == 0) {
+                if (partially_defined_already_warned.count(property.first) == 0) {
+                    warning(
+                        "Extended XYZ",
+                        "property '{}' is only defined for a subset of atoms, it will not be saved",
+                        property.first
+                    );
+                    partially_defined_already_warned.insert(property.first);
+                }
+            }
+        }
+
+        auto to_remove = std::vector<std::string>();
+        for (const auto& property: all_properties) {
+            auto current_property = atom.get(property.first);
+            if (!current_property) {
+                // this property was present for all atoms until now, but
+                // not on the current one
+                warning(
+                    "Extended XYZ",
+                    "property '{}' is only defined for a subset of atoms, it will not be saved",
+                    property.first
+                );
+
+                to_remove.push_back(property.first);
+                continue;
+            }
+
+            if (current_property.value().kind() != property.second) {
+                warning(
+                    "Extended XYZ",
+                    "property '{}' is defined with different types on different atoms, if will not be saved",
+                    property.first
+                );
+                partially_defined_already_warned.insert(property.first);
+                to_remove.push_back(property.first);
+                continue;
+            }
+        }
+
+        for (const auto& name: to_remove) {
+            all_properties.erase(name);
+        }
+    }
+
+    auto results = properties_list_t();
+    for (auto property: std::move(all_properties)) {
+        results.push_back({std::move(property.first), std::move(property.second)});
+    }
```

Comment:
Again, this might be too pedantic but can one not simply return the map `all_properties` instead if iterating over everything?
The only thing that changes is the use of `property.first/second` instead of `property.name/type` in just a few places.

Reply:
I see your concern, but I think I prefer to have clearer code instead of optimizing this too early. I expect that in most cases this would contains 0-10 values, so this iteration is not very costly.

Reply:
Ok, I get the value of having clear code and you're right that one can reasonably expect a small number of properties :+1: